Shipping rates UI changes: Change `saveShippingRates` into custom hook

ecgan opened the pull request:
### Changes proposed in this Pull Request:

Closes #1287.

In this PR, we changed the `saveShippingRates` wp.data actions into custom hook. This addresses the points noted in issue #1287:

1. We add `upsertShippingRates` and `deleteShippingRates` wp.data actions, and use `saveShippingRates` custom hook that wraps around them.
2. `upsertShippingRates` will only be called for changed or newly added shipping rates.
3. We have a `getDeleteIds` function to encapsulate the logic of how to get the IDs to be deleted.
4. No try/catch block in `upsertShippingRates` and `deleteShippingRates` wp.data actions and in `saveShippingRates` custom hook, so that the error can propagate up and be caught by the components to display the necessary error message based on the context.

### Screenshots:

📹 Demo video with my voice:

https://user-images.githubusercontent.com/417342/157846888-e14921e8-2ec3-4ccb-84da-3af1059e9d06.mov


### Detailed test instructions:

You can test this in Setup MC or Edit Free Listings.

1. Open browser dev tools to observe network requests. It is recommended for you to have the Redux Dev Tools to observe wp.data store changes.
2. Go to shipping rate section. Add, edit, or delete shipping rates.
    - Note: to delete shipping rate, you may have to modify the target audience countries.
3. Click on "Continue" / "Save changes" button. In the network panel, there should be a delete shipping rate request for deleted shipping rates, and upsert shipping rate request for added or edited shipping rates. When the requests are done, it should be reflected in the wp.data store.
4. Without reloading the page, repeat step 2 and 3 above for a few more times. This is to test the comparison between the shipping rates from form data and wp.data store.
5. Reload the page. The shipping rates should be the same as before the reload.

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

ecgan committed:
Change saveShippingRates from wp.data action to custom hook.

Also adds upsertShippingRates and deleteShippingRates in wp.data used by saveShippingRates.

ecgan committed:
Replace usage of wp.data saveShippingRates action with the custom hook version.

ecgan committed:
Add error handling in useAutoClearShippingEffect.

This is because saveShippingRates does not have error handling in it by design.

ecgan committed:
Code refactor with getDeleteIds.

ecgan committed:
Change hasUnsavedShippingRates to getUnsavedShippingRates.

This is so that we can use getUnsavedShippingRates later for saveShippingRates operation.

ecgan committed:
Fix getUnsavedShippingRates returning boolean wrongly.

ecgan committed:
Make upsertShippingRates to be called for unsaved shipping rates only.

eason9487 commented on js/src/edit-free-campaign/index.js:
```diff
@@ -148,10 +149,11 @@ export default function EditFreeCampaign() {
 		shippingRates,
 		savedShippingRates
 	);
+	const didRatesChanged = unsavedShippingRates.length > 0;
```
🚧 
`differenceWith` won't tell whether any items in the first array don't exist in the second array. So It's unable to get a correct result of `didRatesChanged` when shipping rates are deleted.

```js
differenceWith( [ 1, 2, 3 ], [ 1, 2, 3 ] );
// > []

differenceWith( [ 1 ], [ 1, 2, 3 ] );
// > []
```

![Kapture 2022-03-14 at 11 32 22](https://user-images.githubusercontent.com/17420811/158102039-43b32c7f-eeae-43db-9d8d-ba327bf67d3c.gif)

ecgan replied:
Thanks. That was my mistake from changing the `hasUnsavedShippingRates` to `getUnsavedShippingRates`, sorry about that. I have reverted the changes there in [`549f26f` (#1313)](https://github.com/woocommerce/google-listings-and-ads/pull/1313/commits/549f26f133191617876642d032fe6afb96d017ca). 🙏

eason9487 commented on js/src/setup-mc/setup-stepper/setup-free-listings/index.js:
```diff
@@ -41,7 +42,8 @@ const SetupFreeListings = ( props ) => {
 	} = useTargetAudienceFinalCountryCodes();
-	const { saveSettings, saveShippingRates } = useAppDispatch();
+	const { saveSettings } = useAppDispatch();
+	const { saveShippingRates } = useSaveShippingRates();
```
💅 

> A further thought is, if that API error would be thrown to this component, then perhaps we could move `setSaving( false )` into `catch` block. The continue button could be set back to enabled only if errors occurred, otherwise, keeping it showing the loading state before bringing merchant to the next step should be fine.

The above is quoted from [my comment in PR 1267](https://github.com/woocommerce/google-listings-and-ads/pull/1267#discussion_r821328150). Sorry for bothering this. As result of the error could be thrown out now, I'd like to mention it again in case of it was just missed.

ecgan replied:
Actually I think that is fine. In case for some reasons the `onContinue` doesn't go to the next page, having the button displayed there with loading state might look like a bug.

Anyway, I'm fine with moving it into the catch block too (because `onContinue` does go to the next page), so I have moved it there in [`43513b9` (#1313)](https://github.com/woocommerce/google-listings-and-ads/pull/1313/commits/43513b9dd139c8aecd30bffc01f0a6caf2c10bad).

eason9487 commented on js/src/setup-mc/setup-stepper/setup-free-listings/useAutoSaveShippingRatesEffect.js:
```diff
@@ -1,12 +1,12 @@
+	const { saveShippingRates } = useSaveShippingRates();
 	const shippingRatesRefValue = useIsEqualRefValue( shippingRates );
 
 	useDebouncedCallbackEffect( shippingRatesRefValue, saveShippingRates );
```
🚧 
The errors could be thrown out from `upsertShippingRates` and `deleteShippingRates` to `saveShippingRates`, but errors are unable to be handled after `saveShippingRates` is wrapped debounce. It may cause an uncaught error when using `useAutoSaveShippingRatesEffect`.

![Kapture 2022-03-14 at 12 31 12](https://user-images.githubusercontent.com/17420811/158105292-62714e15-199c-40f8-90e5-ca5e9a09032e.gif)

The root cause can be simplified as the `bar` function as follow

```js
// Catchable error
async function foo() {
	throw new Error();
}

// Uncatchable error - Simulate the debounce callback
async function bar() {
	setTimeout( () => {
		throw new Error();
	} );
}

( async function () {
	try {
		await foo();
	} catch ( e ) {
		console.log( e );
	}

	try {
		await bar();
	} catch ( e ) {
		console.log( e );
	}
} )();
```

ecgan replied:
Thanks @eason9487. For this auto-save feature, I think we do not need to present a UI prompt to notify users on error, so we can just catch the error and do nothing. I'll work on it.

ecgan replied:
This is fixed in [`bd89578` (#1313)](https://github.com/woocommerce/google-listings-and-ads/pull/1313/commits/bd8957890f9c22e2db4452f660bdef60dbafeed1).

eason9487 replied:
> This is fixed in [`bd89578` (#1313)](https://github.com/woocommerce/google-listings-and-ads/pull/1313/commits/bd8957890f9c22e2db4452f660bdef60dbafeed1).

Sorry, I'm not sure if I was missing something, but looks like this issue may not have been fixed?

https://user-images.githubusercontent.com/17420811/158566092-9d49e7e0-5cc3-4293-9eb6-b0572bb36118.mp4

ecgan replied:
D'oh, it wasn't you who is missing something, it was me missing async await. 🤦‍♂️ Fixed in [`a69a8fd` (#1313)](https://github.com/woocommerce/google-listings-and-ads/pull/1313/commits/a69a8fd30c011e82fa006e2ba56658bae996a93f).

Here's a demo video of the fix:

https://user-images.githubusercontent.com/417342/158599930-9335521e-26fe-47dc-9328-4bc23e6f49bc.mp4

eason9487 commented on js/src/edit-free-campaign/getUnsavedShippingRates.js:
```diff
@@ -18,15 +18,14 @@ import { isEqual, differenceWith, at } from 'lodash';
-		return true;
-	}
-
+export default function getUnsavedShippingRates(
```
❓ 
The original `hasUnsavedShippingRates` also contains the logic to determine whether any deletion has occurred, but the `getUnsavedShippingRates` function only returns the changed shipping rates but does not include those deleted ones. Would it be better to use a name like `getDiffShippingRates` or something like that?

Further, instead of changing the original `hasUnsavedShippingRates` interface, I'd suggest extracting the following parts into a new function. When writing tests, `if ( rates.length ! == savedRates.length )` and `diffRates.length > 0` could be included in test cases together, rather than externalizing these logic to *edit-free-campaign/index.js*.

https://github.com/woocommerce/google-listings-and-ads/blob/34cacd305135b79ba0fda599a8f20db170404eaf/js/src/edit-free-campaign/hasUnsavedShippingRates.js#L30-L40

ecgan replied:
> The original `hasUnsavedShippingRates` also contains the logic to determine whether _any deletion has occurred_, but the `getUnsavedShippingRates` function only returns the changed shipping rates but _does not include those deleted ones_.

🤦 My bad. When I saw the `if ( rates.length ! == savedRates.length )`, somehow I thought it is used to just short-circuit and quickly return a result without having to run `differenceWith` computation. I'll think of something to make things work.

ecgan replied:
I have come up with a new function called `getDifferentShippingRates` in [`5f8649b` (#1313)](https://github.com/woocommerce/google-listings-and-ads/pull/1313/commits/5f8649b99da1ca81795ff1205857e02f8726e6b8). This should now be fixed.

eason9487 commented on js/src/data/reducer.js:
```diff
@@ -124,6 +124,35 @@ const reducer = ( state = DEFAULT_STATE, action ) => {
 			return setIn( state, 'mc.shipping.rates', action.shippingRates );
 		}
 
+		case TYPES.UPSERT_SHIPPING_RATES: {
```
💅 I think maybe it would be better to add back the related reducer tests that were deleted in #1262

ecgan replied:
Oh yes! Done in [`c5b9818` (#1313)](https://github.com/woocommerce/google-listings-and-ads/pull/1313/commits/c5b9818e4a3f7ee919c20ff15e60efac7a8beab0).

jorgemd24 commented on js/src/data/actions.js:
```diff
@@ -75,30 +73,7 @@ export function* fetchShippingRates() {
-}
-
-function* upsertShippingRates( shippingRates ) {
+export function* upsertShippingRates( shippingRates ) {
```
💅

It would be nice to add a JSDoc describing the argument `shippingRates` that is receving this function.

ecgan replied:
Done in [`367c7c9` (#1313)](https://github.com/woocommerce/google-listings-and-ads/pull/1313/commits/367c7c92a512b98a6211e8ea529df4e9830bfc48).

jorgemd24 commented on js/src/data/actions.js:
```diff
@@ -107,57 +82,32 @@ function* upsertShippingRates( shippingRates ) {
-		const upsertedShippingRates = yield upsertShippingRates(
-			newShippingRates
-		);
+export function* deleteShippingRates( ids ) {
```
💅

It would be nice to indicate that ids is an array of ids.

ecgan replied:
Done in [`367c7c9` (#1313)](https://github.com/woocommerce/google-listings-and-ads/pull/1313/commits/367c7c92a512b98a6211e8ea529df4e9830bfc48).

jorgemd24 reviewed (commented):
Thanks, @ecgan 👍  I have left a couple of comments about some functions. In general is working well, except for the things that @eason9487 mentioned before.

ecgan committed:
Revert changes to hasUnsavedShippingRates.

This reverts commit 2da3806088c111c7d4fda16788dde2b447580551, 7eab10a78ff8c6202762fe067a9e971aca9d22d8 and 53058fd8b5e35d7cbd7632733018128c6566ed6f.

ecgan committed:
Add tests for hasUnsavedShippingRates.

ecgan committed:
Code refactor with getDifferentShippingRates.

ecgan committed:
Edit code comment for getDifferentShippingRates.

ecgan committed:
Use getDifferentShippingRates in saveShippingRates.

ecgan committed:
Code refactor for getDeleteIds.

ecgan committed:
setSaving to false only when there is an error.

ecgan committed:
Error handling in useAutoSaveShippingRatesEffect.

ecgan committed:
Add reducer tests for upsert and delete shipping rates.

ecgan committed:
Add JSDoc for upsertShippingRates and deleteShippingRates.

ecgan committed:
Code refactor with getDeletedShippingRates.

eason9487 commented on js/src/edit-free-campaign/hasUnsavedShippingRates.test.js:
```diff
@@ -0,0 +1,258 @@
+ */
+import hasUnsavedShippingRates from './hasUnsavedShippingRates';
+
+describe( 'hasUnsavedShippingRates', () => {
```
💅 

I was going to make up the test for #1307 before, and after comparing some test cases that are missed here, one of the missing cases are partially covered in getDifferentShippingRates.test.js though.

I attach what I wrote as a reference, and the missing parts are marked with a code comment `// Could be added`.

```js
/**
 * External dependencies
 */
import { cloneDeep, set } from 'lodash';

/**
 * Internal dependencies
 */
import { SHIPPING_RATE_METHOD } from '.~/constants';
import hasUnsavedShippingRates from './hasUnsavedShippingRates';

const method = SHIPPING_RATE_METHOD.FLAT_RATE;

describe( 'hasUnsavedShippingRates', () => {
	let savedRates;
	let rates;

	beforeEach( () => {
		savedRates = [
			{
				id: '1',
				country: 'US',
				rate: 12.34,
				currency: 'USD',
				options: { free_shipping_threshold: 1000 },
				method,
			},
			{
				id: '2',
				country: 'JP',
				rate: 4900,
				currency: 'JPY',
				options: { free_shipping_threshold: 2000 },
				method,
			},
		];
		rates = cloneDeep( savedRates );
	} );

	it( 'when both shipping rates are the same, should return false', () => {
		expect( hasUnsavedShippingRates( rates, savedRates ) ).toBe( false );
	} );

	// Could be added
	it( 'when the `id` of shipping rates are different, should still return false', () => {
		delete rates[ 0 ].id;
		delete savedRates[ 1 ].id;

		expect( hasUnsavedShippingRates( rates, savedRates ) ).toBe( false );
	} );

	// Could be added
	it( 'when the `options.free_shipping_threshold` property of one of the shipping rates does not exist and the other is `undefined`, should still return false', () => {
		rates[ 0 ].options.free_shipping_threshold = undefined;
		savedRates[ 0 ].options = {};

		expect( hasUnsavedShippingRates( rates, savedRates ) ).toBe( false );
	} );

	// Could be added
	it( 'the comparison is not sequential, two same sets of shipping rates should return false', () => {
		rates.reverse();

		expect( hasUnsavedShippingRates( rates, savedRates ) ).toBe( false );
	} );

	it( 'when there are deleted shipping rates, should return true', () => {
		rates.pop();

		expect( hasUnsavedShippingRates( rates, savedRates ) ).toBe( true );
	} );

	it( 'when there are new shipping rates, should return true', () => {
		rates.push( {} );

		expect( hasUnsavedShippingRates( rates, savedRates ) ).toBe( true );
	} );

	// Could be added.
	// (Test each property in an atomical way.)
	it.each( [
		[ 'currency', 'AUD' ],
		[ 'rate', 50.88 ],
		[ 'country', 'AU' ],
		[ 'options', {} ],
		[ 'options.free_shipping_threshold', undefined ],
		[ 'options.free_shipping_threshold', 5000 ],
		[ 'method', 'bumpy_rate' ],
	] )(
		'when the property `%s` is edited, for example to %s, should return false',
		( path, value ) => {
			set( rates[ 0 ], path, value );

			expect( hasUnsavedShippingRates( rates, savedRates ) ).toBe( true );
		}
	);
} );
```

ecgan replied:
Your tests there look so good, I ended up using all of them and replacing mine in [`6a9e5f8` (#1313)](https://github.com/woocommerce/google-listings-and-ads/pull/1313/commits/6a9e5f8eac3ef1d652cfabd0d20ca14aa6bf1065). 😆 👍 

(I have my own opinions in testing, but I don't hold them as strong as my other opinions 😆 , and yours are good in here. 😄)

PS: the `(Test each property in an atomical way.)` with `it.each` is nice. 🙂 👍 

Thank you! 🙏

eason9487 commented on js/src/edit-free-campaign/hasUnsavedShippingRates.js:
```diff
@@ -11,6 +11,14 @@ import { isEqual, differenceWith, at } from 'lodash';
  * This function compares the key fields of two shipping rate arrays
  * and returns whether there are any unsaved shipping changes.
  *
+ * There are unsaved shipping changes when there are:
+ *
+ * - deleted shipping rates.
+ * - newly added shipping rates.
+ * - edited shiping rates.
```
Thanks for adding more detailed descriptions! 💯

eason9487 reviewed (commented):
🚧  or 📜 
Found another issue that already existed before this project. When there is a failed API call, the notification at the bottom still shows "Your changes to your Free Listings have been saved and will be synced to your Google Merchant Center account."

https://github.com/woocommerce/google-listings-and-ads/blob/154565b1739ba5a9982caaab60447a05503d00ba/js/src/edit-free-campaign/index.js#L188-L208

https://user-images.githubusercontent.com/17420811/158565390-9cf4b921-0adf-4eda-b37e-a6ec65f972a2.mp4


Maybe this could be fixed in the meantime, or I could open a follow-up issue for it. (I don't really remember, maybe there is already a related issue).

ecgan committed:
Fix missing async await.

ecgan commented:
@eason9487 ,

> Found another issue that already existed before this project. When there is a failed API call, the notification at the bottom still shows "Your changes to your Free Listings have been saved and will be synced to your Google Merchant Center account."
> ...
> Maybe this could be fixed in the meantime, or I could open a follow-up issue for it. (I don't really remember, maybe there is already a related issue).

Yeah it is an existing issue, I noticed it when I was testing it too, thought of fixing it later (too many things for this shipping rate project already). I have created issue https://github.com/woocommerce/google-listings-and-ads/issues/1331 for this, with my thoughts on how to fix it in the issue. 🙏

ecgan committed:
Add and change tests for hasUnsavedShippingRates.

eason9487 commented on js/src/edit-free-campaign/hasUnsavedShippingRates.test.js:
```diff
@@ -0,0 +1,93 @@
+		[ 'options.free_shipping_threshold', 5000 ],
+		[ 'method', 'bumpy_rate' ],
+	] )(
+		'when the property `%s` is edited, for example to %s, should return false',
```
🙏  Sorry, I wrote the wrong expected result here. 🤦‍♂️  
```suggestion
		'when the property `%s` is edited, for example to %s, should return true',
```

ecgan replied:
I was too excited and didn't notice that either 😆 Fixed in [`6624e83` (#1313)](https://github.com/woocommerce/google-listings-and-ads/pull/1313/commits/6624e836019c23039a9bf49fac54f095323ca6b7).

eason9487 commented on js/src/setup-mc/setup-stepper/setup-free-listings/useAutoSaveShippingRatesEffect.js:
```diff
@@ -1,15 +1,37 @@
+	 * We don't want to show error messages for this auto save feature,
+	 * and want it to fail silently in the background.
+	 */
+	const saveShippingRatesCallback = useCallback(
+		async ( value ) => {
+			try {
+				await saveShippingRates( value );
+			} catch ( error ) {}
+		},
+		[ saveShippingRates ]
+	);
+
+	useDebouncedCallbackEffect(
+		shippingRatesRefValue,
+		saveShippingRatesCallback
+	);
```
❔ 
Due to the async call of debounce, it requires wrapping an additional layer of function to catch the error. Considering the problem that the debounce processing was trying to solve already doesn't exist, and the [consistency reason](https://github.com/woocommerce/google-listings-and-ads/pull/1267#discussion_r820441914) I didn't understand, it presents a more unnecessary complexity now.

If it doesn't offer any benefit to maintain consistency with the debounce processing of `useAutoSaveSettingsEffect`, I would still recommend simplifying it.

ecgan replied:
I have created issue https://github.com/woocommerce/google-listings-and-ads/issues/1335 to simplify things even further, by replacing the auto effects with Form `onChange`. We can address the debounce processing there, and by having things together inside `onChange` handler (instead of separate auto effects), I think it should make things easier to understand too. 🙂

eason9487 commented on js/src/setup-mc/setup-stepper/setup-free-listings/useAutoSaveShippingRatesEffect.js:
```diff
@@ -1,15 +1,37 @@
+	 * We don't want to show error messages for this auto save feature,
+	 * and want it to fail silently in the background.
+	 */
+	const saveShippingRatesCallback = useCallback(
+		async ( value ) => {
+			try {
+				await saveShippingRates( value );
+			} catch ( error ) {}
+		},
+		[ saveShippingRates ]
+	);
```
💅 
The callback reference changes don't affect the use of `useDebouncedCallback`. The `useCallback` could be omitted.

ecgan replied:
Ahh right, I forgot that is already handled by [`useDebouncedCallback`](https://github.com/xnimorz/use-debounce#debounced-callbacks). Removed in [`e9f512b` (#1313)](https://github.com/woocommerce/google-listings-and-ads/pull/1313/commits/e9f512bcb99d7baa61ab1760504b8919d4059229). 👍

eason9487 reviewed (approved):
Thanks for the improvement and adding test cases! LGTM. Left some non-blocking comments for minor things.

ecgan committed:
Fix test description for hasUnsavedShippingRates.

Co-authored-by: Eason <eason.su.tw@gmail.com>

ecgan committed:
Remove unneeded useCallback.

We don't need useCallback because that is already handled by useDebouncedCallbackEffect.
